bench/wtperf/wtperf.c

```diff
@@ -2703,6 +2718,45 @@ wtperf_rand(WTPERF_THREAD *thread)
         if (rval > end_range)
             rval = 0;
     }
+
+    /*
+     * A distribution that selects the record with a higher key with higher probability. This was
+     * added to support the YCSB-D workload, which calls for "read latest" selection for records
+     * that are read.
+     *
+     * We generate a random number that is in the range between 0 and largest * largest, where
+     * largest is the last inserted key. Then we take a square root of that random number -- this is
+     * our target selection. With this formula, larger keys are more likely to get selected than
+     * smaller keys, and the probability of selection is proportional to the value of the key, which
+     * is what we want.
+     */
+    if (opts->select_latest) {
+
+        /*
+         * First we need a 64-bit random number, and the WiredTiger random number function gives us
+         * only a 32-bit random value. With only a 32-bit value, the range of the random number will
+         * always be smaller than the square of the largest insert key for workloads with a large
+         * number of keys. So we need a longer random number for that.
+         *
+         * We get a 64-bit random number by concatenating two 32-bit numbers. We get less entropy
+         * this way than via a true 64-bit generator, but we are not defending against cryptographic
+         * attacks here, so this is good enough.
+         */
+        rval2 = __wt_random(&thread->rnd);
+        rval64 = rval | (rval2 << 32);
+
+        /*
+         * Now we limit the random value to be within the range of square of the latest insert key
+         * and take a square root of that value.
+         */
+        rval64 = (rval64 % (wtperf->insert_key * wtperf->insert_key));
+        rval64 = sqrt((long double)rval64);
```

Comment:
If the value of `wtperf-insert_key` is larger than 32 bits, say 0x100000001, won't the expression:
```
(wtperf->insert_key * wtperf->insert_key)
```
overflow 64 bits, and become a small number?  That is, you would get `rval64 % <some_small_number>`

Reply:
@ddanderson, yes, indeed, this can overflow. Good catch! I can change that to use 128-bit types.

Reply:
@ddanderson, could you please take a look at the latest commits?